### PR TITLE
feat(consensus): Add difficulty/pow ethhash checks in consensus pre merge 

### LINKED ISSUE #9476: Add difficulty/pow ethhash checks in consensus pre merge
### Describe the feature

https://github.com/paradigmxyz/reth/blob/d943e782947014a530384c827588cb4db35e86f4/crates/ethereum/consensus/src/lib.rs#L186-L188

## TODO
* add necessary checks

geth ref:
https://github.com/ethereum/go-ethereum/blob/37590b2c5579c36d846c788c70861685b0ea240e/consensus/ethash/consensus.go#L308-L311

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by SozinM**
Closes https://github.com/paradigmxyz/reth/issues/9476

### CODE REVIEW

File: crates/chainspec/src/spec.rs [hunk 1]
```diff
@@ -528,9 +528,12 @@ impl ChainSpec {
                     ..Default::default()
                 }
             }
-            ForkCondition::TTD { total_difficulty, .. } => {
+            ForkCondition::TTD { total_difficulty, fork_block: None } => {
```
Comment (SozinM) [inline crates/chainspec/src/spec.rs:531]: left this to keep backward compatible behaviour

**Comment by SozinM** [comment]
@mattsse minibump :)

File: crates/chainspec/src/spec.rs [hunk 2]
```diff
@@ -1333,15 +1357,19 @@ Post-merge hard forks (timestamp based):
...
                 ),
                 (
                     EthereumHardfork::Shanghai,
-                    ForkId { hash: ForkHash([0xdc, 0xe9, 0x6c, 0x2d]), next: 1710338135 },
+                    ForkId { hash: ForkHash([0xa7, 0x01, 0xb4, 0xc9]), next: 1710338135 },
                 ),
                 (
                     EthereumHardfork::Cancun,
-                    ForkId { hash: ForkHash([0x9f, 0x3d, 0x22, 0x54]), next: 0 },
+                    ForkId { hash: ForkHash([0x74, 0x9e, 0x0c, 0xdc]), next: 0 },
```
Comment (onbjerg) [inline crates/chainspec/src/spec.rs:1372]: these fork ids should not change. they are used in discovery, so its important they stay fixed to line up with what every other client agrees on

File: crates/consensus/common/src/difficulty.rs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+## Algorithm
+`block_diff = pdiff+pdiff/2048 * (1 if time-ptime < 13 else -1) + int(2^((num//100000) - 2))`
+
+Where:
+- `pdiff  = parent.difficulty`
+- `ptime = parent.time`
+- `time = block.timestamp`
+- `num = block.number`
+**/
```
Comment (onbjerg) [inline crates/consensus/common/src/difficulty.rs:72]: please use `/// ` as is standard in rust

File: crates/consensus/common/src/difficulty.rs [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        let period_count =
+            U256::from(1).checked_shl(exponent).ok_or(CalculatorError::OverflowError)?;
+        pdiff += period_count;
+    }
+    // block_diff = pdiff + pdiff / 2048 * (1 if time - ptime < 13 else -1) + int(2^((num // 100000)
+    // - 2))
+    Ok(pdiff)
+}
+
+/**
```
Comment (onbjerg) [inline crates/consensus/common/src/difficulty.rs:104]: same as above

File: crates/consensus/common/src/difficulty.rs [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+        let exponent =
+            exponent.saturating_sub(2).try_into().map_err(|_| CalculatorError::ConversionError)?;
+        let period_count =
+            U256::from(1).checked_shl(exponent).ok_or(CalculatorError::OverflowError)?;
+        pdiff += period_count;
+    }
+
+    Ok(pdiff)
+}
+/**
```
Comment (onbjerg) [inline crates/consensus/common/src/difficulty.rs:168]: same as above

File: crates/consensus/common/src/difficulty.rs [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    Eip2384 = 9_000_000,
+    /// Specification EIP-3554: <https://eips.ethereum.org/EIPS/eip-3554>
+    Eip3554 = 9_700_000,
+    /// Specification EIP-4345: <https://eips.ethereum.org/EIPS/eip-4345>
+    Eip4345 = 10_700_000,
+    /// Specification EIP-5133: <https://eips.ethereum.org/EIPS/eip-5133>
+    Eip5133 = 11_400_000,
+}
+
+// Helper to convert BombDelay to u64
```
Comment (onbjerg) [inline crates/consensus/common/src/difficulty.rs:52]: this comment is not necessary

**Review by onbjerg — CHANGES_REQUESTED** [review]
some preliminary comments, sorry for the delay

**Comment by mattsse** [comment]
I noticed that geth will phase out difficulty entirely, effectively deprecating this.

So I think we also don't need to add it, especially with upcoming changes to sync

**Pull request closed without merging**